ui/library/src/components/outline/Outline.tsx

```diff
@@ -1,50 +1,43 @@
 import * as React from 'react';
 
-import { DocumentContext } from '../../context/DocumentContext';
-import { TransformContext } from '../../context/TransformContext';
-import { scrollToPosition } from '../../utils/scroll';
-import { Ref } from '../types/destination';
+import { buildOutlinePositions, DocumentContext } from '../../context/DocumentContext';
+import { ScrollContext } from '../../context/ScrollContext';
 import { NodeDestination, OutlineNode } from '../types/outline';
 import { OutlineItem } from './OutlineItem';
 
 export const Outline: React.FunctionComponent = ({ ...extraProps }) => {
-  const { outline, pdfDocProxy, setOutline } = React.useContext(DocumentContext);
-  const { rotation } = React.useContext(TransformContext);
+  const { outline, pdfDocProxy, setOutline, setOutlinePositions } =
+    React.useContext(DocumentContext);
+  const { scrollToOutlineTarget, resetScrollObservers } = React.useContext(ScrollContext);
 
   if (!pdfDocProxy) {
     return null;
   }
 
-  if (!outline) {
-    pdfDocProxy.getOutline().then((outlineArray: Array<OutlineNode>) => {
-      setOutline(outlineArray);
-    });
-  }
+  React.useEffect(() => {
+    if (outline) {
+      return;
+    }
+    pdfDocProxy
+      .getOutline()
+      .then((outlineArray: Array<OutlineNode>) => {
+        setOutline(outlineArray);
+        return buildOutlinePositions(pdfDocProxy, outlineArray);
+      })
+      .then(outlinePositions => {
+        setOutlinePositions(outlinePositions);
+      })
+      .then(() => {
+        resetScrollObservers();
+      });
```

Comment:
Tbh I am not sure how this 3 lines of code run. The previous callback in `.then` seem not returning any promise unless I missed something here